activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -463,6 +458,22 @@ def schedule_query(future_result) # :nodoc:
       end
 
       private
+        def build_async_executor
+          case Base.async_query_executor
+          when :multi_thread_pool
+            Concurrent::ThreadPoolExecutor.new(
+              min_threads: @db_config.min_threads,
+              max_threads: @db_config.max_threads,
+              max_queue: @db_config.max_queue,
+              fallback_policy: :caller_runs
+            )
+          when :global_thread_pool
+            Base.global_thread_pool_async_query_executor
+          else
+            Base.immediate_query_executor
```

Comment:
I think that using the immediate executor for now is fine, but we might want to actually make this a new NullExecutor object in the future. Right now all the `load_async` behavior is manual and not automatic. But I'm thinking that when we implement automatic async behavior in something like the preloader it might not make sense for applications to use an `ImmediateExecutor` and maybe instead it should use an internal no-op object that knows what to do. I think this is something we can solve in a followup PR though.

Reply:
I was thinking the same yes. Immediate executor will trick the Adapter into thinking it was called from a background thread, so the query will end up logged as `ASYNC` even though it was directly executed in the foreground. This as too much potential to trick users.

Reply:
I'll work on a different kind of object on Monday and think more about how to better define sensible defaults for applications.